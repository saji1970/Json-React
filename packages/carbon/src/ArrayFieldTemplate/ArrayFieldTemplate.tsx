import { Stack } from '@carbon/react';
import {
  getTemplate,
  getUiOptions,
  ArrayFieldTemplateItemType,
  ArrayFieldTemplateProps,
  StrictRJSFSchema,
  RJSFSchema,
  FormContextType,
} from '@rjsf/utils';
import { LayerBackground } from '../components/Layer';
import getCarbonOptions from '../utils';

/** Implement `ArrayFieldTemplate`
 */
export default function ArrayFieldTemplate<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any
>(props: ArrayFieldTemplateProps<T, S, F>) {
  const {
    className,
    canAdd,
    disabled,
    idSchema,
    uiSchema,
    items,
    onAddClick,
    readonly,
    registry,
    required,
    schema,
    title,
  } = props;
  const uiOptions = getUiOptions<T, S, F>(uiSchema);
  const carbonOptions = getCarbonOptions<T, S, F>(registry.formContext, uiOptions);
  const ArrayFieldDescriptionTemplate = getTemplate<'ArrayFieldDescriptionTemplate', T, S, F>(
    'ArrayFieldDescriptionTemplate',
    registry,
    uiOptions
  );
  const ArrayFieldItemTemplate = getTemplate<'ArrayFieldItemTemplate', T, S, F>(
    'ArrayFieldItemTemplate',
    registry,
    uiOptions
  );
  const ArrayFieldTitleTemplate = getTemplate<'ArrayFieldTitleTemplate', T, S, F>(
    'ArrayFieldTitleTemplate',
    registry,
    uiOptions
  );
  // Button templates are not overridden in the uiSchema
  const {
    ButtonTemplates: { AddButton },
  } = registry.templates;
  return (
    <div className={className} id={idSchema.$id}>
      <ArrayFieldTitleTemplate
        idSchema={idSchema}
        title={uiOptions.title || title}
        schema={schema}
        uiSchema={uiSchema}
        required={required}
        registry={registry}
      />
      <ArrayFieldDescriptionTemplate
        idSchema={idSchema}
        description={uiOptions.description || schema.description}
        schema={schema}
        uiSchema={uiSchema}
        registry={registry}
      />
      <LayerBackground padding={carbonOptions.padding}>
        <Stack gap={carbonOptions.gap}>
          {items.length > 0 && (
            <Stack gap={carbonOptions.gap} className='array-item-list'>
              {items.map(({ key, ...itemProps }: ArrayFieldTemplateItemType<T, S, F>) => (
                <ArrayFieldItemTemplate key={key} {...itemProps} />
              ))}
            </Stack>
          )}
          {canAdd && (
            <AddButton
              className='array-item-add'
              onClick={onAddClick}
              disabled={disabled || readonly}
              uiSchema={uiSchema}
              registry={registry}
            />
          )}
        </Stack>
      </LayerBackground>
    </div>
  );
}
