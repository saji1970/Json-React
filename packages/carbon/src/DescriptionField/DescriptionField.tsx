import { DescriptionFieldProps, FormContextType, RJSFSchema, StrictRJSFSchema } from '@rjsf/utils';

/** Implement `DescriptionField`
 */
export default function DescriptionField<
  T = any,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = any
>({ description, id }: DescriptionFieldProps<T, S, F>) {
  if (!description) {
    return null;
  }

  return (
    <div id={id} className='field-description cds--form__helper-text'>
      {description}
    </div>
  );
}
