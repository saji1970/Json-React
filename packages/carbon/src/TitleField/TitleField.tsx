import { FormContextType, RJSFSchema, StrictRJSFSchema, TitleFieldProps } from '@rjsf/utils';
import { LabelValue } from '../components/LabelValue';
import { FormGroup } from '@carbon/react';

/** Implement `TitleFieldTemplate`
 */
export default function TitleField<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>({
  id,
  title,
  required,
}: TitleFieldProps<T, S, F>) {
  return (
    <div className='title-field'>
      <FormGroup legendId={id} legendText={<LabelValue label={title} required={required} hide={false} />}>
        {null}
      </FormGroup>
    </div>
  );
}
