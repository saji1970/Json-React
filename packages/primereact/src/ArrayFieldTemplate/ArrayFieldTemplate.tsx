import React from "react";
import { utils } from "@rjsf/core";
import { ArrayFieldTemplateProps, IdSchema } from "@rjsf/core";

import IconButton from "../IconButton/IconButton";

const { isMultiSelect, getDefaultRegistry } = utils;

const ArrayFieldTemplate = (props: ArrayFieldTemplateProps) => {
  const { schema, registry = getDefaultRegistry() } = props;

  if (isMultiSelect(schema, registry.rootSchema)) {
    return <DefaultFixedArrayFieldTemplate {...props} />;
  }
  return <DefaultNormalArrayFieldTemplate {...props} />;
};

type ArrayFieldTitleProps = {
  TitleField: any;
  idSchema: IdSchema;
  title: string;
  required: boolean;
};

const ArrayFieldTitle = ({
  TitleField,
  idSchema,
  title,
  required,
}: ArrayFieldTitleProps) => {
  if (!title) {
    return null;
  }

  const id = `${idSchema.$id}__title`;
  return (
    <TitleField id={id} title={title} required={required} />
  );
};

type ArrayFieldDescriptionProps = {
  DescriptionField: any;
  idSchema: IdSchema;
  description: string;
};

const ArrayFieldDescription = ({
  DescriptionField,
  idSchema,
  description,
}: ArrayFieldDescriptionProps) => {
  if (!description) {
    return null;
  }

  const id = `${idSchema.$id}__description`;
  return (
    <DescriptionField id={id} description={description} />
  );
};

// Used in the two templates
const DefaultArrayItem = (props: any) => {
  const canMoveItems = props.hasMoveUp || props.hasMoveDown;
  return (
    <div key={props.key} className="flex align-items-start gap-2 p-2 border-1 border-round">
      <div className="flex-grow-1">{props.children}</div>
      <div>
        {props.hasToolbar && (
          <div className="flex flex-row">
            {canMoveItems && (
              <>
                <IconButton
                  icon="arrow-up"
                  className="array-item-move-up"
                  style={props.hasRemove ? {
                    borderTopRightRadius: 0,
                    borderBottomRightRadius: 0,
                  } : undefined}
                  disabled={props.disabled || props.readonly || !props.hasMoveUp}
                  onClick={props.onReorderClick(props.index, props.index - 1)}
                />
                <IconButton
                  icon="arrow-down"
                  style={{
                    borderLeft: 0,
                    borderTopLeftRadius: 0,
                    borderBottomLeftRadius: 0,
                    ...(props.hasRemove ? {
                      borderTopRightRadius: 0,
                      borderBottomRightRadius: 0,
                    } : {}),
                  }}
                  disabled={props.disabled || props.readonly || !props.hasMoveDown}
                  onClick={props.onReorderClick(props.index, props.index + 1)}
                />
              </>
            )}

            {props.hasRemove && (
              <IconButton
                icon="remove"
                style={canMoveItems ? {
                  borderLeft: 0,
                  borderTopLeftRadius: 0,
                  borderBottomLeftRadius: 0,
                } : undefined}
                disabled={props.disabled || props.readonly}
                onClick={props.onDropIndexClick(props.index)}
              />
            )}
          </div>
        )}
      </div>
    </div>
  );
};

const DefaultFixedArrayFieldTemplate = (props: ArrayFieldTemplateProps) => {
  return (
    <fieldset className={props.className}>
      <ArrayFieldTitle
        key={`array-field-title-${props.idSchema.$id}`}
        TitleField={props.TitleField}
        idSchema={props.idSchema}
        title={props.uiSchema["ui:title"] || props.title}
        required={props.required}
      />

      {(props.uiSchema["ui:description"] || props.schema.description) && (
        <div
          className="field-description"
          key={`field-description-${props.idSchema.$id}`}>
          {props.uiSchema["ui:description"] || props.schema.description}
        </div>
      )}

      <div
        className="array-item-list flex flex-column gap-2"
        key={`array-item-list-${props.idSchema.$id}`}
      >
        {props.items && props.items.map(DefaultArrayItem)}
      </div>

      {props.canAdd && (
        <IconButton
          icon="plus"
          className="mt-3 mb-3 array-item-add"
          onClick={props.onAddClick}
          disabled={props.disabled || props.readonly}
        />
      )}
    </fieldset>
  );
};

const DefaultNormalArrayFieldTemplate = (props: ArrayFieldTemplateProps) => {
  return (
    <>
      <ArrayFieldTitle
        key={`array-field-title-${props.idSchema.$id}`}
        TitleField={props.TitleField}
        idSchema={props.idSchema}
        title={props.uiSchema["ui:title"] || props.title}
        required={props.required}
      />

      {(props.uiSchema["ui:description"] || props.schema.description) && (
        <ArrayFieldDescription
          key={`array-field-description-${props.idSchema.$id}`}
          DescriptionField={props.DescriptionField}
          idSchema={props.idSchema}
          description={
            props.uiSchema["ui:description"] || props.schema.description
          }
        />
      )}

      <div key={`array-item-list-${props.idSchema.$id}`} className="flex flex-column gap-2">
        {props.items && props.items.map(p => DefaultArrayItem(p))}

        {props.canAdd && (
          <IconButton
            icon="plus"
            className="mt-1 mb-3 array-item-add"
            onClick={props.onAddClick}
            disabled={props.disabled || props.readonly}
          />
        )}
      </div>
    </>
  );
};

export default ArrayFieldTemplate;
