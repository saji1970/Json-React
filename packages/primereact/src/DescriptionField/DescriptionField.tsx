import React from "react";
import { FieldProps } from "@rjsf/core";

const DescriptionField = ({ description }: Partial<FieldProps>) => {
  if (description) {
    return <div className="text-sm mt-1 mb-3">{description}</div>;
  }

  return null;
};

export default DescriptionField;
