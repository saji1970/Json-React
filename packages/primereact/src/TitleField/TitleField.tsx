import React from "react";
import { FieldProps } from "@rjsf/core";

const TitleField = ({ title, uiSchema }: Pick<FieldProps, "title" | "uiSchema">) => (
  <div className="border-bottom-1 mb-2">
    <h5>{(uiSchema && uiSchema["ui:title"]) || title}</h5>
  </div>
);

export default TitleField;
