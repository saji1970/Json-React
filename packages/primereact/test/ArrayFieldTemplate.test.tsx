import React from "react";
import ArrayFieldTemplate from "../src/ArrayFieldTemplate";
import DescriptionField from "../src/DescriptionField";
import renderer from "react-test-renderer";
import TitleField from "../src/TitleField";
import { mockSchema } from "./helpers/createMocks";
import { utils } from "@rjsf/core";
const { getDefaultRegistry } = utils;

describe("ArrayFieldTemplate", () => {
  test("simple", () => {
    const tree = renderer
      .create(
        <ArrayFieldTemplate
          DescriptionField={() => <DescriptionField description={null} />}
          TitleField={() => <TitleField uiSchema={{}} />}
          items={[]}
          canAdd={true}
          className=""
          disabled={false}
          onAddClick={() => void 0}
          readonly={false}
          required={false}
          schema={mockSchema}
          uiSchema={{}}
          title=""
          formContext={{}}
          formData={{}}
          registry={{ ...getDefaultRegistry() }}
          // @ts-ignore
          idSchema={{}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
