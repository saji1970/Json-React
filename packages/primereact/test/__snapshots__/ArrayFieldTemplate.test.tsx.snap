// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayFieldTemplate simple 1`] = `
<div
  className="flex flex-column gap-2"
>
  <button
    className="p-button p-component p-button-outlined p-button-sm mt-1 mb-3 array-item-add p-button-icon-only"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <i
      className="pi pi-plus"
    />
    <span
      className="p-button-label p-c"
      dangerouslySetInnerHTML={
        Object {
          "__html": "&nbsp;",
        }
      }
    />
  </button>
</div>
`;
